### PR TITLE
Added accounts microservice

### LINKED ISSUE #23: Add authentication logic
**Is your feature request related to a problem? Please describe.**
No

**Describe the solution you'd like**
Add some authentication logic that would work with the login page (#22)

**Describe alternatives you've considered**
N/A

**Additional context**
N/A

### LINKED ISSUE #32: Account management and login microservice
This was missing for some reason??

### PR DESCRIPTION
**Pull request by Duszke333**
Closes #32 Closes #23 

# Description
An implementation of the account management and logging in microservice.
Fully documented using Swagger.
Tests included.

# API endpoints:
- `GET /accounts/` - retrieves all existing accounts information
 - `POST /accounts/` - creates an account
 - `GET /accounts/{user_id}` - retrieves information about an user
 - `PUT /accounts/{user_id}` - updates information about an user
 - `DELETE /accounts/{user_id}` - deletes an user from the database
 - `GET /login/` - try to log in user (check if password hash is correct and get user information if so)
 - `GET /login/{username}` - get the password salt associated with the user (for hashing the password before logging in)
 
# The login procedure:
1. User types in their username and password
2. Frontend asks the microservice for the password salt (`GET /login/{username}`)
3. Frontend hashes the password using the salt
4. Frontend sends username and hashed password for authorization ( `GET /login/`)

### CODE REVIEW

File: accounts/db/main_test.go
Comment (maksnowak) [inline accounts/db/main_test.go:1]: Do we really need tests like this? It's bloat IMO

File: accounts/docs/main_test.go
Comment (maksnowak) [inline accounts/docs/main_test.go:1]: Ditto

File: accounts/docs/swagger.yaml [hunk 1]
```diff
@@ -1,4 +1,130 @@
...
+  models.HTTP500:
+    properties:
+      <plain string>:
+        example: Error while processing request
+        type: string
+    type: object
+  models.Register:
+    properties:
+      email:
+        example: huan.pablo.dos@vatican.city
```
Comment (maksnowak) [inline accounts/docs/swagger.yaml:39]: 😵‍💫

File: accounts/docs/swagger.yaml [hunk 2]
```diff
@@ -1,4 +1,130 @@
...
+        type: string
+      passwordHash:
+        example: Kremowki
```
Comment (maksnowak) [inline accounts/docs/swagger.yaml:42]: 😵‍💫

File: accounts/docs/swagger.yaml [hunk 3]
```diff
@@ -1,4 +1,130 @@
...
+        type: string
+      passwordSalt:
+        example: Slony_Karmel
```
Comment (maksnowak) [inline accounts/docs/swagger.yaml:45]: 😵‍💫

File: accounts/docs/swagger.yaml [hunk 4]
```diff
@@ -1,4 +1,130 @@
...
+        type: string
+      role:
+        example: admin
+        type: string
+      username:
+        example: Karol_Wojtyla
```
Comment (maksnowak) [inline accounts/docs/swagger.yaml:51]: 😵‍💫

File: accounts/docs/swagger.yaml [hunk 5]
```diff
@@ -1,4 +1,130 @@
...
+  models.Update:
+    properties:
+      email:
+        example: huan.pablo.tres@vatican.city
+        type: string
+      ownedFiles:
+        example:
+        - rower
+        - pies
+        - zachrystia
```
Comment (maksnowak) [inline accounts/docs/swagger.yaml:72]: 😵‍💫

File: accounts/docs/swagger.yaml [hunk 6]
```diff
@@ -1,4 +1,130 @@
...
+      passwordHash:
+        example: Papiezowki
+        type: string
+      passwordSalt:
+        example: Pozdrawiam_Polakow
+        type: string
+      sharedFiles:
+        example:
+        - zaba
+        - cialo_chrystusa
```
Comment (maksnowak) [inline accounts/docs/swagger.yaml:85]: 😵‍💫

File: accounts/docs/swagger.yaml [hunk 7]
```diff
@@ -1,4 +1,130 @@
...
+        - pies
+        items:
+          type: string
+        type: array
+      passwordHash:
+        example: Kremowki
+        type: string
+      passwordSalt:
+        example: Slony_Karmel
+        type: string
```
Comment (maksnowak) [inline accounts/docs/swagger.yaml:113]: 😵‍💫

File: accounts/docs/swagger.yaml [hunk 8]
```diff
@@ -1,4 +1,130 @@
...
+      sharedFiles:
+        example:
+        - zaba
+        - slon
+        items:
+          type: string
+        type: array
+      username:
+        example: Karol_Wojtyla
```
Comment (maksnowak) [inline accounts/docs/swagger.yaml:125]: 😵‍💫

File: accounts/main_test.go [hunk 9]
```diff
@@ -2,15 +2,6 @@ package main
...
-	}
-}
-
-func TestAddOne(t *testing.T) {
-	x := 7
-	if addOne(&x) != 8 {
-		t.Fatal("7 + 1 does not equal 8")
-	}
+func TestAll(t *testing.T) {
+	t.Log("Setting up the router and connecting to the database, nothing to test here")
```
Comment (maksnowak) [inline accounts/main_test.go:6]: Ditto (unnecessary test files)

File: accounts/models/main_test.go [hunk 10]
```diff
@@ -0,0 +1,7 @@
+package models
+
+import "testing"
+
+func TestAll(t *testing.T) {
+	t.Log("Just plain data structures, nothing to test here")
```
Comment (maksnowak) [inline accounts/models/main_test.go:6]: Ditto

File: accounts/models/user.go [hunk 11]
```diff
@@ -0,0 +1,59 @@
...
+
+// # User
+//
+// A structure representing MongoDB User structure
+type User struct {
+	ID           bson.ObjectID `bson:"_id,omitempty" json:"id" example:"675f9a97ca1d148373316ae4"`
+	Username     string        `bson:"username" json:"username" example:"Karol_Wojtyla"`
+	Email        string        `bson:"email" json:"email" example:"huan.pablo.dos@vatican.city"`
+	PasswordHash string        `bson:"passwordHash" json:"passwordHash" example:"Kremowki"`
+	PasswordSalt string        `bson:"passwordSalt" json:"passwordSalt" example:"Slony_Karmel"`
```
Comment (maksnowak) [inline accounts/models/user.go:16]: 😵‍💫

File: accounts/models/user.go [hunk 12]
```diff
@@ -0,0 +1,59 @@
...
+	OwnedFiles   []string      `bson:"ownedFiles" json:"ownedFiles" example:"rower,pies"`
+	SharedFiles  []string      `bson:"sharedFiles" json:"sharedFiles" example:"zaba,slon"`
+}
+
+// # Salt
+//
+// A structure containing the username and password salt associated with it
+type Salt struct {
+	Username     string `bson:"username" json:"username" example:"Karol_Wojtyla"`
+	PasswordSalt string `bson:"passwordSalt" json:"passwordSalt" example:"Slony_Karmel"`
```
Comment (maksnowak) [inline accounts/models/user.go:28]: 😵‍💫

File: accounts/models/user.go [hunk 13]
```diff
@@ -0,0 +1,59 @@
...
+}
+
+// # Register
+//
+// A structure representing data needed to register an account
+type Register struct {
+	Username     string `bson:"username" json:"username" example:"Karol_Wojtyla"`
+	Email        string `bson:"email" json:"email" example:"huan.pablo.dos@vatican.city"`
+	PasswordHash string `bson:"passwordHash" json:"passwordHash" example:"Kremowki"`
+	PasswordSalt string `bson:"passwordSalt" json:"passwordSalt" example:"Slony_Karmel"`
```
Comment (maksnowak) [inline accounts/models/user.go:46]: 😵‍💫

File: accounts/models/user.go [hunk 14]
```diff
@@ -0,0 +1,59 @@
...
+
+// # Update
+//
+// A structure representing account data that can be updated
+type Update struct {
+	Email        string   `bson:"email" json:"email" example:"huan.pablo.tres@vatican.city"`
+	PasswordHash string   `bson:"passwordHash" json:"passwordHash" example:"Papiezowki"`
+	PasswordSalt string   `bson:"passwordSalt" json:"passwordSalt" example:"Pozdrawiam_Polakow"`
+	OwnedFiles   []string `bson:"ownedFiles" json:"ownedFiles" example:"rower,pies,zachrystia"`
+	SharedFiles  []string `bson:"sharedFiles" json:"sharedFiles" example:"zaba,cialo_chrystusa"`
```
Comment (maksnowak) [inline accounts/models/user.go:58]: 😵‍💫

**Review by maksnowak — CHANGES_REQUESTED** [review]
Please change the examples for something else, also I would remove the empty test files, they really are unnecessary as they provide no code coverage anyways

File: accounts/db/main_test.go
Comment (Duszke333) [reply accounts/db/main_test.go:1]: Without it, the statistics of test coverage are wrong (for example, in models without a test it shows 0%, with it there is an information that there are no statements present). Also, this way you can gain information about the packages when you ran tests with -v flag (verbose).

Comment (maksnowak) [reply accounts/db/main_test.go:1]: Okay, in this case the test can stay, I don't know enough about Go to be the judge (@Depermitto your thoughts?)

**Comment by maksnowak** [comment]
Ready to merge

**Pull request merged**